### PR TITLE
Ability to use custom query string manager

### PR DESCRIPTION
**Pull request by igieon**
- documentation
- tests

### CODE REVIEW

File: tests/test_sqlalchemy_data_layer.py [hunk 1]
```diff
@@ -410,6 +421,29 @@ class PersonList(ResourceList):
...
+        def _simple_filters(self, dict_):
+            return [{"name": key, "op": "in" if isinstance(value, list) else "eq", "val": value}
+                    for (key, value) in dict_.items()]
+
+    class PersonList(ResourceList):
+        schema = person_schema
+        data_layer = {
+            "model": person_model,
+            "session": session,
+            "methods": {"before_create_object": before_create_object},
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py]: `methods` is not used and `before_create_object` won't be overridden, you can remove this line

File: tests/test_sqlalchemy_data_layer.py [hunk 2]
```diff
@@ -767,14 +803,41 @@ def test_get_list_with_simple_filter(client, register_routes, person, person_2):
         querystring = urlencode(
             {
                 "page[number]": 1,
-                "page[size]": 1,
+                "page[size]": 10,
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py]: why?

File: tests/test_sqlalchemy_data_layer.py [hunk 3]
```diff
@@ -767,14 +803,41 @@ def test_get_list_with_simple_filter(client, register_routes, person, person_2):
...
                 "fields[person]": "name,birth_date",
                 "sort": "-name",
                 "filter[name]": "test",
             }
         )
         response = client.get("/persons" + "?" + querystring, content_type="application/vnd.api+json")
         assert response.status_code == 200
+        assert len(response.json['data']) == 1
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py]: will it be `1` always if page size is changed (above)?

File: tests/test_sqlalchemy_data_layer.py [hunk 4]
```diff
@@ -767,14 +803,41 @@ def test_get_list_with_simple_filter(client, register_routes, person, person_2):
...
+    with client:
+        querystring = urlencode(
+            {
+                "filter[computers.id]": f'{computer_2.id},{computer.id}',
+                "include": "computers"
+            }
+        )
+        response = client.get("/persons_qs" + "?" + querystring, content_type="application/vnd.api+json")
+        assert response.status_code == 200
+        assert len(response.json['data']) == 2
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py:859]: maybe also to check that ids are as expected?

File: tests/test_sqlalchemy_data_layer.py [hunk 5]
```diff
@@ -767,14 +803,41 @@ def test_get_list_with_simple_filter(client, register_routes, person, person_2):
...
+
+        querystring = urlencode(
+            {
+                "filter[computers.id]": f'{computer_2.id},{computer.id}',
+                "include": "computers"
+            }
+        )
+        response = client.get("/persons" + "?" + querystring, content_type="application/vnd.api+json")
+        assert response.status_code == 500
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py]: this should be moved to a separate test. tests should be atomic. you can move repeating parts to fixtures

**Review by mahenzon — CHANGES_REQUESTED** [review]
Nice! Cool feature, but tests need some minor updates

File: tests/test_sqlalchemy_data_layer.py (hunk 1 above)
Comment (igieon) [reply tests/test_sqlalchemy_data_layer.py]: i removed it

File: tests/test_sqlalchemy_data_layer.py (hunk 2 above)
Comment (igieon) [reply tests/test_sqlalchemy_data_layer.py]: I returned back and check meta count

File: tests/test_sqlalchemy_data_layer.py (hunk 4 above)
Comment (igieon) [reply tests/test_sqlalchemy_data_layer.py:859]: check ids

File: tests/test_sqlalchemy_data_layer.py (hunk 3 above)
Comment (igieon) [reply tests/test_sqlalchemy_data_layer.py]: changed to meta count

File: tests/test_sqlalchemy_data_layer.py [hunk 6]
```diff
@@ -158,7 +158,7 @@ def engine(
         person_tag_model, person_single_tag_model, person_model,
         computer_model, string_json_attribute_person_model,
         address_model
-    ):
```
Comment (mahenzon) [inline tests/test_sqlalchemy_data_layer.py:161]: Oops

**Review by mahenzon — APPROVED** [review]
Ok

File: tests/test_sqlalchemy_data_layer.py (hunk 6 above)
Comment (igieon) [reply tests/test_sqlalchemy_data_layer.py:161]: Do I need rollback this whitespace error ?
I use idea ultimate with automatic formatter so probably how its get there. Could by nice to have some formatter definition like flake8 or pylint.

Comment (mahenzon) [reply tests/test_sqlalchemy_data_layer.py:161]: Well, no need
Yeah, we better add another pipeline / git hook with formatting

**Pull request merged**